src/Microsoft.ML.StandardTrainers/Standard/SdcaBinary.cs

```diff
@@ -802,6 +802,7 @@ private void InitializeConvergenceMetrics(out string[] names, out Double[] initi
                 Func<DataViewRowId, long> getIndexFromId = GetIndexFromIdGetter(idToIdx, biasReg.Length);
                 while (cursor.MoveNext())
                 {
+                    Host.CheckAlive();
```

Comment:
This looks good - would it be worth adding a couple of tests that cancel the training?

Reply:
Yea we have unit-tests that test cancellation signal has been set correctly. For real-time testing it will require much more work. I still need to figure if that is something we do with a test or manually test it. All in all the current mechanism for cancellation is temporary, long term goal is to use cancellation token. I will highly encourage you read the issue associated with this PR and comment on my analysis.

Reply:
Thanks Zeeshan.

As for analysis - I think where you have the cancellation check is reasonable. Even looking through the SDCA code, this looks like the two places where we loop and you would want to check the cancellation condition. I did not see any additional places where this needs to be checked. 

As for testsing, I don't think the test would be much more work  - you can start training an SDCA pipeline, call cancel on the environment (CancelExecution I believe) and then check if the OperationCancelledExecption is thrown. Since Train is a blocking call, you would have to call it in a separate thread -- but I believe that is already expected since we don't have any TrainAsync APIs.

I do have a couple additional questions:
1) CheckAlive is ifdef'ed by a !CPUMATH_INFRASTRUCTURE -- however CPUMATH_INFRASTRUCTURE is defined in Microsoft.ML.CpuMath - how is CheckAlive enabled?
2) CheckAlive throws an OperationCancelledException - is ml.net supposed to catch that exception or is this up to the user to catch?

Reply:
Its not as simple, there are all sorts of timing issues. I'm manually testing this with a customer and once I validate this works in practice then I will check-in as it is. The user catches the exception.